packages/downloading-helpers/src/file-downloads/download-file.ts

```diff
@@ -1,29 +1,79 @@
-import { createWriteStream } from "fs";
+import { createWriteStream, existsSync } from "fs";
 import { rm } from "fs/promises";
-import { finished } from "stream";
+import { Stream, finished } from "stream";
 import { promisify } from "util";
 import axios from "axios";
 import axiosRetry from "axios-retry";
 import extract from "extract-zip";
 
 const promisifiedFinished = promisify(finished);
 
-export const downloadFile: (
+export const downloadFile = async (
   fileUrl: string,
-  path: string
-) => Promise<void> = async (fileUrl, path) => {
+  path: string,
+  opts: {
+    firstDataTimeoutInMs?: number;
+    downloadCompleteTimeoutInMs?: number;
+    maxDownloadContentRetries?: number;
+    downloadContentRetryDelay?: number;
+  } = {}
+): Promise<void> => {
   // Using the same timeout as the standard client in meticulous-sdk/packages/client/src/client.ts
-  const client = axios.create({ timeout: 60_000 });
+  const firstDataTimeoutInMs = opts.firstDataTimeoutInMs ?? 60_000;
+  const downloadCompleteTimeoutInMs =
+    opts.downloadCompleteTimeoutInMs ?? 120_000;
+  const maxDownloadContentRetries = opts.maxDownloadContentRetries ?? 3;
+  const downloadContentRetryDelay = opts.downloadContentRetryDelay ?? 1000;
+
+  const client = axios.create({ timeout: firstDataTimeoutInMs });
   axiosRetry(client, { retries: 3, shouldResetTimeout: true });
+  const source = axios.CancelToken.source();
+
+  const response = await client.request({
+    method: "GET",
+    url: fileUrl,
+    responseType: "stream",
+    cancelToken: source.token,
+  });
 
   const writer = createWriteStream(path);
+  (response.data as Stream).pipe(writer);
+  const timeoutId = setTimeout(async () => {
+    const error = `Download timed out after ${downloadCompleteTimeoutInMs}ms`;
+    source.cancel(error);
+    writer.destroy(new Error(error));
+  }, downloadCompleteTimeoutInMs);
+
+  try {
+    await promisifiedFinished(writer);
+    if (timeoutId) {
+      clearTimeout(timeoutId);
+    }
+  } catch (err) {
+    if (timeoutId) {
+      clearTimeout(timeoutId);
+    }
 
-  return client
-    .request({ method: "GET", url: fileUrl, responseType: "stream" })
-    .then(async (response) => {
-      response.data.pipe(writer);
-      return promisifiedFinished(writer);
+    await new Promise((resolve) => writer.close(resolve));
+
+    if (existsSync(path)) {
+      await rm(path);
```

Comment:
Could we document that this function isn't safe to call concurrently from different processes (i.e. it should not be called within replays and only at the test run level)? I had tried to make it safe to do this, but reverted that because it broke stuff, and now with this line it's perhaps even less safe.

Reply:
Ah yeah that's true. I'll have a think and can document it. My assumption here was that if we've started writing to the file and failed mid way through it's probably already corrupted, so better to just delete it.

Reply:
Alternatively we could revive https://github.com/alwaysmeticulous/meticulous-sdk/pull/646 but fix it. The issue was `fileSync` from `tmp` returned a file path in the system's temporary directory but on cloud replay nodes that was on a different file system to the directory we were downloading into so `renameSync` would throw.